src/ringmesh/basic/plugin_manager.cpp

```diff
@@ -0,0 +1,266 @@
+                    FORMAT_MESSAGE_FROM_SYSTEM
+                    // allocate buffer on local heap for error text
+                    |FORMAT_MESSAGE_ALLOCATE_BUFFER
+                    // Important! will fail otherwise, since we're not
+                    // (and CANNOT) pass insertion parameters
+                    |FORMAT_MESSAGE_IGNORE_INSERTS,
+                    NULL,// unused with FORMAT_MESSAGE_FROM_SYSTEM
+                    GetLastError(),
+                    MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+                    (LPTSTR) &errorText,// output
+                    0,// minimum size for output buffer
+                    NULL);// arguments - see note
+
+                if( errorText != nullptr )
+                {
+                    message = errorText;
+                    LocalFree(errorText);
+                }
+                throw RINGMeshException( "Plugin", "Could not load ", plugin_path,
+                    ": ", message );
+            }
+        }
+
+        std::string home_directory() const
+        {
+            PWSTR path{ nullptr };
+            HRESULT hr = SHGetKnownFolderPath(&FOLDERID_Documents, 0, nullptr, &path);
+            std::string result;
+            if( SUCCEEDED( hr ) )
+            {
+                result = path;
+            }
+            CoTaskMemFree(path);
+            return result;
+        }
+
+        std::string executable_directory() const
+        {
+            TCHAR buff[MAX_PATH];
+            GetModuleFileName( NULL, buff, MAX_PATH);
+            return GEO::FileSystem::dir_name( std::string{ buff } );
+        }
+
+    private:
+        std::string library_name( const std::string& plugin_name ) const
+        {
+            return { plugin_name + ".dll" };
+        }
+    };
+} // namespace RINGMesh
+#else
+#include <dlfcn.h>
+#include <pwd.h>
+
+namespace RINGMesh
+{
+    class PluginManager::Impl
+    {
+    public:
+        /*!
+         * Loads the given library.
+         * RTLD_NOW: All undefined symbols in the library are resolved before dlopen() returns.
+         *          If this cannot be done, an error is returned.
+         * RTLD_GLOBAL: The symbols defined by this library will be made available
+         *          for symbol resolution of subsequently loaded libraries.
+         */
+        void load_library( const std::string& plugin_name ) const
+        {
+            std::string library{ library_name( plugin_name ) };
+            void* plugin_handle = dlopen( library.c_str(),
+            RTLD_NOW | RTLD_GLOBAL );
+            if( plugin_handle == nullptr )
+            {
+                throw RINGMeshException( "Plugin", "Could not load ", library,
+                    ": ", dlerror() );
+            }
+        }
+
+        std::string home_directory() const
+        {
+            std::string homedir{ getenv( "HOME" ) };
+            if( homedir.empty() )
+            {
+                homedir = getpwuid( getuid() )->pw_dir;
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Replace this call to the non reentrant function 'getpwuid' by a call to 'getpwuid_r'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3ANonReentrantFunction)